### ISSUE TITLE
Webpack generated library triggers Error: "Angular JIT compilation failed: '@angular/compiler' not loaded!"

### ISSUE THREAD

**Issue by mprinc** [angular/angular#44436]
### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

No

### Description

We have a regular and working web component built from an NG project. We use it in another hosting app (vue app in this current case).

All works fine, the web component uses a config file `global.js` which is injected into global space inside the `public/index.html` file.

However, if we compile the config file from the original `global.ts` file through webpack we have the abovementioned error.

If we remove the webpack bootstrap parts all works fine again.

Any understanding of the cause of the problem and how to resolve it? Thaks.

This is the webpack bootstrap heading part we removed:

```js
/******/ (() => { // webpackBootstrap
/******/ 	"use strict";
/******/ 	// The require scope
/******/ 	var __webpack_require__ = {};
/******/
/************************************************************************/
/******/ 	/* webpack/runtime/make namespace object */
/******/ 	(() => {
/******/ 		// define __esModule on exports
/******/ 		__webpack_require__.r = (exports) => {
/******/ 			if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 				Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 			}
/******/ 			Object.defineProperty(exports, '__esModule', { value: true });
/******/ 		};
/******/ 	})();
/******/
/************************************************************************/
var __webpack_exports__ = {};
// ESM COMPAT FLAG
__webpack_require__.r(__webpack_exports__);

;// CONCATENATED MODULE: foo/bar.ts
;
```

and webpack bootstrap footter part:

```js
window.global = __webpack_exports__;
/******/ })()
;
//# sourceMappingURL=data:application/json;charset=utf-8;base64,
// ... the rest of the map
````

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 10.0.6
Node: 14.18.1
OS: darwin x64

Angular: 10.2.5
... animations, common, compiler, compiler-cli, core, elements
... forms, material, platform-browser, platform-browser-dynamic
... platform-server, router, service-worker
Ivy Workspace: <error>

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.1000.6
@angular-devkit/build-angular      0.1000.6
@angular-devkit/build-ng-packagr   0.1000.6
@angular-devkit/build-optimizer    0.1000.6
@angular-devkit/build-webpack      0.1000.6
@angular-devkit/core               10.0.6
@angular-devkit/schematics         10.0.6
@angular/cdk                       10.2.7
@angular/cli                       10.0.6
@angular/flex-layout               10.0.0-beta.32
@angular/http                      8.0.0-beta.10+1.sha-a28b3e3
@angular/language-service          10.0.6
@schematics/angular                <error>
@schematics/update                 <error>
rxjs                               6.6.2
```

**Comment by JoostK** [comment]
Libraries have historically been built using the ViewEngine compiler, which was Angular's render engine until Angular 9. Since then, ViewEngine libraries need to be processed to become with Angular Ivy, the renderer engine since v9. This is done using `ngcc` (Angular Compatibility Compiler) which has to statically analyze the JS bundles of built libraries to insert the necessary code to work with Ivy. Since this relies on static analysis, only well known bundle formats are supported: ESM, UMD and CommonJS. The Webpack envelope that you have will therefore cause the static analysis of this bundle not to find anything.

I don't know where that Webpack envelope comes from, perhaps you can add it _after_ ngcc has processed the bundle. Alternatively, Angular 12 and newer support native distribution of Ivy packages, such that ngcc is no longer needed.

**Comment by mprinc** [comment]
Hi @JoostK,

Well the "problematic" file has nothing with angular or has to be processed with ngcc at all, or it even can/should be, as it is not a library, it is a mere config file that ends up injecting its content into the window global var and becomes available to the rest of the NG world.

Its natural state is to be changed, evolved and rebuilt **after** the rest of the NG components are already built, so there is no static analysis. With the pure config file (basically pure JS) there is no problem with the rest of the NG world (real-time problem, as there is no static analysis of such file anyway; it is not known to NG built pipeline by its content, but rather just used at the runtime, after being injected). However, the webapck bootstrap triggers the problem in the realtime, which is neither expected nor its analyses is needed. 

I do not even understand why the webpack bootstrap would trigger such a runtime behavior of angular, when pure file doesn't do that?

**Comment by mprinc** [comment]
Primarily, I am wondering how I can avoid this unpleasant scenario where, on every change of the config file (it is originally a relatively complex Typescript file, as it has config interfaces, etc),  I have to recompile it with webpack and then manually (automatically) clean part of legitimate code to avoid NG complaining about missing '@angular/compiler' and breaking the whole app because of it.

**Comment by JoostK** [comment]
Can you check whether the compiler was also needed when the Webpack envelope was not present? If you set a breakpoint in the `getCompilerFacade` function within `@angular/core`; if that function is invoked then you can be certain that the JIT compiler is used.

It might be that the Webpack envelope is clearing earlier module registrations, especially since there is things like `window.global = __webpack_exports__;` in there (which affects the global scope). I'm only guessing here, but it might give a pointer.

**Comment by mprinc** [comment]
Yes, here it is, called **many times**, for code without Webpack envelope

![image](https://user-images.githubusercontent.com/996001/146281307-301115f8-b1c7-45cd-9836-3368d70fd954.png)

and it continued fine after that.

If I leave the envelope, it is called once and crashes just after that:

![image](https://user-images.githubusercontent.com/996001/146282268-4b844784-2005-46d1-8cc5-ff5c89d6424b.png)


And the whole web component crashes and it is not recognized anymore in the hosting app

**Comment by JoostK** [comment]
Thanks for the update. That indicates that your app has not been AOT compiled, triggering compilation in the browser using the JIT compiler. I don't know why the Webpack envelope causes the compiler instance to be lost, I suspect that it interferes with other Webpack stuff.

**Comment by mprinc** [comment]
Correct, but again there is no additional webpack stuff except angular web
component being packed with webpack by angular-cli (which is mostly out of
my scope). On the other hand, the config file is regularly built as a
webpack library and therefore it seems highly unlikely that it would have
a problem with other (ng web component) webpack instances.

Again, I feel it should be in angular arena to resolve this problem, as
webpack cannot be blamed for it. I would understand if webpack is a 3rd
party tool, but correct me if I'm wrong, no matter how deeply inside
"hidden", webpack IS used for building angular? So there is nothing wrong
to expect that regular webpack-workflow and bundling: having two separate
web bundles that work together in the browser is a breaking scenario?

**Comment by JoostK** [comment]
If you post a reproduction I can have a look as to why the compiler gets lost.

FYI, the usage of Webpack for bundling within the CLI is an implementation detail, it may change in the future.

**Comment by mprinc** [comment]
Thanks for the offer I will try, but again, I need to simplify and provide
very slim example of a complex solution, web component in angular, vue host
app (which again would be hard, I assume, to dwell in, and) and external
config files.
I will put reminder in Feb, and until then will automatize the webpack
cleaning procedure. Thanks again.

For many reasons I am excited with Snowpack, but again Webpack is much more
mature and will most likely imitate all good things from snowpack anyway :)
